### ISSUE TITLE
The process cannot access the file .poshssh\hosts.json because it is being used by another process

### ISSUE THREAD

**Issue by okarpov** [darkoperator/Posh-SSH#494]
System.Management.Automation.ActionPreferenceStopException: The running command stopped because the preference variable "ErrorActionPreference" or common parameter is set to Stop: The process cannot access the file 'C:\Users\user1\.poshssh\hosts.json' because it is being used by another process.
   at System.Management.Automation.Runspaces.PipelineBase.Invoke(IEnumerable input)
   at System.Management.Automation.PowerShell.Worker.ConstructPipelineAndDoWork(Runspace rs, Boolean performSyncInvoke)
   at System.Management.Automation.PowerShell.Worker.CreateRunspaceIfNeededAndDoWork(Runspace rsToUse, Boolean isSync)
   at System.Management.Automation.PowerShell.CoreInvokeHelper[TInput,TOutput](PSDataCollection`1 input, PSDataCollection`1 output, PSInvocationSettings settings)
   at System.Management.Automation.PowerShell.CoreInvoke[TInput,TOutput](PSDataCollection`1 input, PSDataCollection`1 output, PSInvocationSettings settings)
   at System.Management.Automation.PowerShell.Invoke[T](IEnumerable input, IList`1 output)
   at Scripter.Form1.DownloadFile(PSSLog psslog) in E:\user1\Projects\Scripter\Form1.cs:line 906
   at Scripter.Form1.<>c__DisplayClass26_1.<btnRun_Click>b__1() in E:\user1\Projects\Scripter\Form1.cs:line 480

**Comment by darkoperator** [comment]
Need more context outside of pasting an error. The file seems locked is the only thing I can tell but no clue on what version of windows, what version of PS, version of module or how the command is being ran

**Comment by MVKozlov** [comment]
Seems it like a #457 problem.

Race condition still exists when writing concurrently because `FileStream` with `Share.Read` do not try to wait on fail.

I don't know if some work is needed or if a workaround is enough.

**Comment by okarpov** [comment]
@darkoperator  simplified version attached. 
( I thought devs know how, when and why their module uses its own .poshssh\hosts.json file and thus it would be obviouse. )

**Comment by MVKozlov** [comment]
@okarpov, You, of course, also can look at issue above, the code of the module and draw conclusions :)

Hint: pre-populate `hosts.json` with keys before using it so it's never written to while your code is running. Or use one-time in memory key storage if you don't care about keys.

**Comment by darkoperator** [comment]
As a dev it is the minimum I expect in a proper issue is to provide context with details on how to reproduce :)

**Comment by darkoperator** [comment]
Not part of the current problem but something you might hit is this issue with the library https://github.com/sshnet/SSH.NET/pull/844 for a very long time there has been many issues in the library with threading that have not been resolved.
